README
Description:
A simple KDE Plasmoid front-end to the GeeXboX uShare UPnP media server utility.

Starts a uShare instance with the list of shared folders. Created it with the primary purpose of being able to easily share my video/music files with my Xbox360.

Requirements:
KDE4 >= 4.x
GeeXboX uShare: <http://ushare.geexbox.org/>

Source
Current source available at git repository:
<http://github.com/amoskahiga/ushareoid>

Features:
- Add folders to share.
- Start and stop the uShare instance.
- Set common uShares startup options/settings.

Installation:
Extract contents and run installation scripts:
- install_local.sh will install to the local users profile
- install_global.sh will install to the global profile and requires root privileges to install

Alternatively you can run following commands:
>mkdir build
>cd build
>cmake ../ DCMAKE_INSTALL_PREFIX=~/.kde4
>make
>make install
>kbuildsycoca4

After this, the Plasmoid should show up in the add widgets dialog.

CHANGELOG
0.2 (20.May.2009)
- Added optional settings for applet.

0.1 (16.May.2009)
- Initial version.

applet/ushareoid.h
/***************************************************************************
 *   Copyright (C) 2009 Amos Kariuki    <amoskahiga@hotmail.com>           *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA .        *
 ***************************************************************************/

#ifndef USHAREOID_H
#define USHAREOID_H

#include <Plasma/PopupApplet>
#include <QProcess>

#include "configdialog.h"

class QGraphicsGridLayout;
class QStringListModel;
class QGraphicsWidget;
class KTempDir;

namespace Plasma
{
    class PushButton;
    class TextEdit;
    class TreeView;
    class IconWidget;
    class Label;
}

/**
 * A plasmoid that provides a front-end to the popular GeeXboX uShare utility. It starts a uShare
 * instance with a list of shared folders.
 *
 * @see Hosted at ushareoid: <http://www.kde-look.org/content/show.php?content=105025>
 * @see Git repository: <http://github.com/amoskahiga/ushareoid>
 */
class Ushareoid : public Plasma::PopupApplet
{
    Q_OBJECT

    private:

        QGraphicsGridLayout* m_layout;
        QGraphicsWidget *m_widget;
        QStringListModel* m_folderListModel;
        QProcess* m_process;
        KTempDir* m_virtualFolder;
        Plasma::Label* m_sharedLabel;
        Plasma::TreeView* m_folderView;
        Plasma::PushButton* m_removeFolderButton;
        Plasma::PushButton* m_addFolderButton;
        Plasma::PushButton* m_shareButton;
        Plasma::TextEdit* m_statusEdit;
        Plasma::IconWidget* m_icon;
        ConfigDialog* m_configDialog;
        ConfigDialog::Settings m_settings;
        enum {NOT_SHARING, SHARING} m_sharingStatus;

        virtual QGraphicsWidget *graphicsWidget();
        void createConfigurationInterface(KConfigDialog* parent);
        QString createVirtualFolder(QStringList folderList);

    public:

        // Basic Create/Destroy
        Ushareoid(QObject *parent, const QVariantList &args);
        ~Ushareoid();

    private slots:

        void addFolderButtonClick();
        void removeFolderButtonClick();
        void shareButtonClick();
        void updateStatusEdit();
        void processStarted();
        void processFinished(int,QProcess::ExitStatus);
        void toolTipAboutToShow();
        void configAccepted();

};

// This is command links the applet to the .desktop file
K_EXPORT_PLASMA_APPLET(ushareoid, Ushareoid)

#endif // USHAREOID_H


applet/configdialog.h
/***************************************************************************
 *   Copyright (C) 2009 by Amos Kariuki <amoskahiga@hotmail.com>           *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA .        *
 ***************************************************************************/

#ifndef CONFIGDIALOG_H
#define CONFIGDIALOG_H

#include "ui_configdialog.h"

/**
 * A configuration widget containing the settings for the Ushareoid applet.  This class serves as
 * a simple view for the settings.
 */
class ConfigDialog : public QWidget

{
    Q_OBJECT

    public:

        struct Settings
        {
            QString ushareExecutable;
            QString friendlyName;
            QString networkInterface;
            int port;
            bool enableXboxCompliantProfile;
            bool enableWebPageControl;
            bool enableTelnetControl;
         };

        ConfigDialog(QWidget* parent = 0);
        ~ConfigDialog();
        void setSettings(Settings newSettings);
        Settings getSettings();

    private:
        Ui::ConfigDialog configUi;
};

#endif // CONFIGDIALOG_H


applet/configdialog.cpp
/***************************************************************************
 *   Copyright (C) 2009 Amos Kariuki    <amoskahiga@hotmail.com>           *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA .        *
 ***************************************************************************/

#include "configdialog.h"

/**
 * Constructor that build and initialize the widgets UI.
 */
ConfigDialog::ConfigDialog(QWidget *parent)
        : QWidget(parent)
{
    configUi.setupUi(this);
}

/**
 * Destructor.
 */
ConfigDialog::~ConfigDialog()
{

}

/**
 * Gets the settings currently set in the UI.
 * @return Current configuration values held by the UI elements.
 */
ConfigDialog::Settings ConfigDialog::getSettings()
{
    Settings settings;

    settings.ushareExecutable = configUi.ushareExecutableKurlrequester->text().trimmed();
    settings.friendlyName = configUi.friendlyNameLineEdit->text().trimmed();
    settings.networkInterface = configUi.networkInterfaceLineEdit->text().trimmed();
    settings.port = configUi.portSpinBox->value();
    settings.enableXboxCompliantProfile = configUi.enableXboxCompliantProfileCheckBox->isChecked();
    settings.enableTelnetControl = configUi.enableTelnetControlCheckBox->isChecked();
    settings.enableWebPageControl = configUi.enableWebPageControlCheckBox->isChecked();

    return settings;
}

/**
 * Initializes the UI elements with the given values.
 * @param newSettings New configuration to display.  The old values as discarded.
 */
void ConfigDialog::setSettings(Settings newSettings)
{
    configUi.ushareExecutableKurlrequester->setUrl(newSettings.ushareExecutable);
    configUi.friendlyNameLineEdit->setText(newSettings.friendlyName);
    configUi.networkInterfaceLineEdit->setText(newSettings.networkInterface);
    configUi.portSpinBox->setValue(newSettings.port);
    configUi.enableXboxCompliantProfileCheckBox->setChecked(newSettings.enableXboxCompliantProfile);
    configUi.enableTelnetControlCheckBox->setChecked(newSettings.enableTelnetControl);
    configUi.enableWebPageControlCheckBox->setChecked(newSettings.enableWebPageControl);
}


applet/ushareoid.cpp
/***************************************************************************
 *   Copyright (C) 2009 Amos Kariuki    <amoskahiga@hotmail.com>           *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA .        *
 ***************************************************************************/

#include <QGraphicsGridLayout>
#include <QTreeView>
#include <QStringListModel>
#include <QHeaderView>
#include <KUrl>
#include <KDirSelectDialog>
#include <KPushButton>
#include <KTextEdit>
#include <KConfigDialog>
#include <KTempDir>
#include <kio/copyjob.h>
#include <Plasma/ToolTipManager>
#include <Plasma/TreeView>
#include <Plasma/TextEdit>
#include <Plasma/PushButton>
#include <Plasma/Label>

#include "ushareoid.h"

/**
 * Constructor.
 * @param parent Owner.
 * @param args
 */
Ushareoid::Ushareoid(QObject *parent, const QVariantList &args)
    : Plasma::PopupApplet(parent, args),
    m_widget(0),
    m_virtualFolder(0),
    m_configDialog(0),
    m_sharingStatus(NOT_SHARING)
{
    // this will get us the standard applet background, for free!
    setBackgroundHints(DefaultBackground);

    setAspectRatioMode(Plasma::IgnoreAspectRatio);
    setPopupIcon("ushareoid");

    m_process = new QProcess(this);
    m_process->setProcessChannelMode(QProcess::MergedChannels);
    connect(m_process, SIGNAL(readyReadStandardOutput()), this, SLOT(updateStatusEdit()));
    connect(m_process, SIGNAL(started()), this, SLOT(processStarted()));
    connect(m_process, SIGNAL(finished(int,QProcess::ExitStatus)), this,
            SLOT(processFinished(int,QProcess::ExitStatus)));
}

/**
 * Destructor.
 */
Ushareoid::~Ushareoid()
{
    if (hasFailedToLaunch()) {
        // Do some cleanup here
    } else {
        // Save settings
        KConfigGroup configGroup = globalConfig();
        configGroup.writePathEntry("defaultFolders", m_folderListModel->stringList());
    }

    delete m_process;
    delete m_virtualFolder;
}

/**
 * Gets the widget that will get shown in the applet or in a Dialog, depending on the form factor of
 * the applet (it may be docked in a panel or displayed on a desktop).
 */
QGraphicsWidget* Ushareoid::graphicsWidget()
{
    if (!m_widget) {

        // Get the configuration settings
        KConfigGroup configGroup = globalConfig();
        QStringList folderList = configGroup.readPathEntry("defaultFolders", QStringList());
        m_settings.ushareExecutable = configGroup.readEntry("ushareExecutable", "ushare");
        m_settings.friendlyName = configGroup.readEntry("friendlyName", "uShareoid");
        m_settings.networkInterface = configGroup.readEntry("networkInterface", "");
        m_settings.port = configGroup.readEntry("port", 0);
        m_settings.enableXboxCompliantProfile =
                configGroup.readEntry("enableXboxCompliantProfile", true);
        m_settings.enableWebPageControl = configGroup.readEntry("enableWebPageControl", false);
        m_settings.enableTelnetControl = configGroup.readEntry("enableTelnetControl", false);

        // Set default ushare executable name if none is set.
        if(m_settings.ushareExecutable.isEmpty())
            m_settings.ushareExecutable = "ushare";

        setAspectRatioMode(Plasma::IgnoreAspectRatio);

        // Register taskbar tooltips
        Plasma::ToolTipManager::self()->registerWidget(this);

        m_widget = new QGraphicsWidget(this);
        m_layout = new QGraphicsGridLayout(m_widget);

        m_folderListModel = new QStringListModel(m_widget);
        m_folderListModel->setStringList(folderList);

        /**
         * To change the header for the m_folderListModel (Displayed in the TreeView), we would
         * have to subclass QStringListModel and reimplement setHeaderData()
         *
         * @see <http://www.qtsoftware.com/developer/task-tracker/index_html?id=188749&method=entry>

        qDebug() << m_folderListModel->setHeaderData( 0, Qt::Horizontal, "Shared Folders",
                                                      Qt::DisplayRole);
        */

        m_folderView = new Plasma::TreeView(m_widget);
        m_folderView->setModel(m_folderListModel);
        m_folderView->nativeWidget()->header()->hide();
        m_folderView->setStyleSheet("border-color: white; border-width: 1px; border-style: solid;"
                                    "border-radius: 4px;");

        m_sharedLabel = new Plasma::Label(this);
        m_sharedLabel->setText("Shared Folders:");

        m_statusEdit = new Plasma::TextEdit(m_widget);
        m_statusEdit->nativeWidget()->setReadOnly(true);
        m_statusEdit->hide();

        m_addFolderButton = new Plasma::PushButton(m_widget);
        m_addFolderButton->nativeWidget()->setIcon(KIcon("list-add"));
        m_addFolderButton->setText("Add Folder...");
        connect(m_addFolderButton, SIGNAL(clicked()), this, SLOT(addFolderButtonClick()));

        m_removeFolderButton = new Plasma::PushButton(m_widget);
        m_removeFolderButton->nativeWidget()->setIcon(KIcon("list-remove"));
        m_removeFolderButton->setText("Remove Folder");
        connect(m_removeFolderButton, SIGNAL(clicked()), this, SLOT(removeFolderButtonClick()));

        m_shareButton = new Plasma::PushButton(m_widget);
        m_shareButton->nativeWidget()->setIcon(KIcon("media-playback-start"));
        m_shareButton->setText("Start Sharing");
        connect(m_shareButton, SIGNAL(clicked()), this, SLOT(shareButtonClick()));

        // Add widgets to the layout.
        m_layout->addItem(m_sharedLabel, 0, 0);
        m_layout->addItem(m_folderView, 1, 0, 4, 1);
        //m_layout->addItem(m_statusEdit, 3, 0);
        m_layout->addItem(m_addFolderButton, 1, 1);
        m_layout->addItem(m_removeFolderButton, 2, 1);
        m_layout->addItem(m_shareButton, 4, 1, Qt::AlignBottom);

        m_widget->setPreferredSize(500, 300);
    }

    return m_widget;
}

/**
 * Displays a file dialog allowing the user to specify the folder to add.  The selected folder is
 * added to the folder lists.
 */
void Ushareoid::addFolderButtonClick()
{
    KDirSelectDialog dialog(KUrl(), true);

    if (dialog.exec()) {
        QString path = dialog.url().path();
        if (!m_folderListModel->stringList().contains(path)) {
            QStringList list = m_folderListModel->stringList();
            list.append(path);
            list.sort();
            m_folderListModel->setStringList(list);
        }
    }

    // Keep showing the applet--would otherwise minimize if placed in a panel.
    this->showPopup();
}

/**
 * Removes the currently selected folder item.
 */
void Ushareoid::removeFolderButtonClick()
{
    // Determine the selected folder item.
    QString selectedFolder = m_folderView->nativeWidget()->currentIndex().data().toString();

    // Remove the folder from the list.
    QStringList list = m_folderListModel->stringList();
    list.removeAll(selectedFolder);
    m_folderListModel->setStringList(list);
    m_statusEdit->nativeWidget()->insertPlainText(m_process->readAllStandardOutput());
}

/**
 * Starts a "ushare" process with the loaded settings.  The folders listed in the string-list view
 *  will be shared.
 */
void Ushareoid::shareButtonClick()
{
    if (m_process->state() == QProcess::Running) {

        m_process->close();
    }
    else {

        // Set the default executable name if it's not set.
        if (m_settings.ushareExecutable.isEmpty())
            m_settings.ushareExecutable = "ushare";

        // Set application/process arguments based on possible persisted settings.
        QStringList arguments;

        if (!m_settings.friendlyName.isEmpty())
            arguments << "-n" << m_settings.friendlyName;
        else    // Set the default name that shows up in the UPnP browser.
            arguments << "-n" << "uShareoid";

        if (!m_settings.networkInterface.isEmpty())
            arguments << "-i" << m_settings.networkInterface;

        if (m_settings.port)
            arguments << "-p" << QString::number(m_settings.port);

        if (!m_settings.enableTelnetControl)
            arguments << "-t";

        if (!m_settings.enableWebPageControl)
            arguments << "-w";

        if (m_settings.enableXboxCompliantProfile) {
            arguments << "-x";

            /**
             * Generate a new folder-list with all shared items under a new temporary root folder
             * and add the temporary root shared folder to the list of arguments
             */
            arguments << "-c" << createVirtualFolder(m_folderListModel->stringList());
        }
        else
        {
            // Add each folder separately to the list of arguments.
            foreach (QString dir, m_folderListModel->stringList())
                arguments << "-c" << dir;
        }

        m_process->start(m_settings.ushareExecutable, arguments);
    }
}

/**
 * Updates the Status view with the output of the ushare process.  All messages to the stdout or
 * stderr will be written to the view.
 *
 * @todo Does not currently work as intended--the ushare utlitly seems to be buffering the output
 *       thereby preventing the applet from displaying it.
 *       NOTE: Even though the output is immediately displayed when 'ushare' is run from the
 *             terminal, the C runtime library may be detecting the output device and determining
 *             the stdout buffering behaviour.  stderr is not buffered and is available immediately.
 */
void Ushareoid::updateStatusEdit()
{
    //m_statusEdit->nativeWidget()->insertPlainText(m_process->readAll());
    qDebug() << m_process->readAll();
}

/**
 * Toggles the state of the Share button to indicate that uShare is running.
 */
void Ushareoid::processStarted()
{
    m_sharingStatus = SHARING;
    m_shareButton->setText("Stop Sharing");
    m_shareButton->nativeWidget()->setIcon(KIcon("media-playback-stop"));
}

/**
 * Toggles the state of the Share button to indicate that Ushare has ended.
 */
void Ushareoid::processFinished(int exitCode, QProcess::ExitStatus)
{
    Q_UNUSED(exitCode);

    m_sharingStatus = NOT_SHARING;
    Plasma::ToolTipManager::self()->show(this); // Inform the user we've stopped sharing
    m_shareButton->setText("Start Sharing");
    m_shareButton->nativeWidget()->setIcon(KIcon("media-playback-start"));
}

/**
 * Determine and set the sharing status to be displayed in the tooltip.
 */
void Ushareoid::toolTipAboutToShow()
{
    Plasma::ToolTipContent data;
    data.setImage(KIcon("ushareoid").pixmap(IconSize(KIconLoader::Panel)));
    data.setMainText("uShare");

    if (m_sharingStatus == SHARING) {
         data.setSubText("Sharing...");
    }
    else {   // NOT_SHARING
        data.setSubText("Not Sharing...");
    }
    Plasma::ToolTipManager::self()->setContent(this, data);;
}

/**
 * Creates a configuration dialog that will be used to display and set our applet's persisted
 * settings.  The dialog is initialized with our previously persisted settings.
 */
void Ushareoid::createConfigurationInterface(KConfigDialog* parent)
{
    m_configDialog = new ConfigDialog(parent);
    m_configDialog->setSettings(m_settings);
    parent->addPage(m_configDialog, QString("uShare Settings"), icon());
    connect(parent, SIGNAL(okClicked()), this, SLOT(configAccepted()));
}

/**
 * Called when the user chooses to save the changes of the applet's configuration/settings dialog.
 * It persists the settings to the applet's configuration file.
 */
void Ushareoid::configAccepted()
{
    // Get the new settings
    m_settings = m_configDialog->getSettings();

    // Save the configuration settings
    KConfigGroup configGroup = globalConfig();
    configGroup.writeEntry("ushareExecutable", m_settings.ushareExecutable);
    configGroup.writeEntry("friendlyName", m_settings.friendlyName);
    configGroup.writeEntry("networkInterface", m_settings.networkInterface);
    configGroup.writeEntry("port", m_settings.port);
    configGroup.writeEntry("enableXboxCompliantProfile", m_settings.enableXboxCompliantProfile);
    configGroup.writeEntry("enableWebPageControl", m_settings.enableWebPageControl);
    configGroup.writeEntry("enableTelnetControl", m_settings.enableTelnetControl);
}

/**
 * Creates a temporary virtual folder (if it doesn't exist yet) and links in all shared folders.
 * @param folderList list of folders to link into the virtual folder
 */
QString Ushareoid::createVirtualFolder(QStringList folderList)
{
    // If we had already set up a shared virtual folder, delete it and its contents
   delete m_virtualFolder;

   m_virtualFolder = new KTempDir();

    // Link all folders in the list to the virtual folder
    KUrl::List urlList(folderList);
    KIO::link(urlList, KUrl(m_virtualFolder->name()));

    return m_virtualFolder->name();
}

#include "ushareoid.moc"
